test/common/router/router_test.cc

```diff
@@ -1115,6 +1116,81 @@ TEST_F(RouterTest, UpstreamTimeout) {
   EXPECT_TRUE(verifyHostUpstreamStats(0, 1));
 }
 
+// Verify the timeout budget histograms are filled out correctly when using a
+// global and per-try timeout.
+TEST_F(RouterTest, TimeoutBudgetHistogramStat) {
+  NiceMock<Http::MockStreamEncoder> encoder;
+  Http::StreamDecoder* response_decoder = nullptr;
+  EXPECT_CALL(cm_.conn_pool_, newStream(_, _))
+      .WillOnce(Invoke([&](Http::StreamDecoder& decoder, Http::ConnectionPool::Callbacks& callbacks)
+                           -> Http::ConnectionPool::Cancellable* {
+        response_decoder = &decoder;
+        callbacks.onPoolReady(encoder, cm_.conn_pool_.host_, upstream_stream_info_);
+        return nullptr;
+      }));
+  expectPerTryTimerCreate();
+  expectResponseTimerCreate();
+
+  Http::TestHeaderMapImpl headers{{"x-envoy-upstream-rq-timeout-ms", "400"},
+                                  {"x-envoy-upstream-rq-per-try-timeout-ms", "200"}};
+  HttpTestUtility::addDefaultHeaders(headers);
+  router_.decodeHeaders(headers, false);
+  Buffer::OwnedImpl data;
+  router_.decodeData(data, true);
+
+  // These are coming from CodeStatsImpl::chargeResponseTiming.
+  EXPECT_CALL(cm_.thread_local_cluster_.cluster_.info_->stats_store_,
+              deliverHistogramToSinks(_, 80ull))
+      .Times(3);
+  // Global timeout budget used.
+  EXPECT_CALL(cm_.thread_local_cluster_.cluster_.info_->stats_store_,
+              deliverHistogramToSinks(_, 20ull));
```

Comment:
Can you be explicit about which histogram is getting which value in these tests. I think you can do something like:
https://github.com/envoyproxy/envoy/blob/master/test/extensions/filters/network/redis_proxy/command_splitter_impl_test.cc#L602